checker/cost.go

```diff
@@ -0,0 +1,536 @@
+package checker
+
+import (
+	"fmt"
+	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"math"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/containers"
+	"github.com/google/cel-go/common/types/ref"
+)
+
+// CostEstimator estimates the sizes of variable length input data and the costs of functions.
+type CostEstimator interface {
+	// EstimateSize returns the worst case size possible for the given Expr, or nil if
+	// the estimator has no estimate to provide. EstimateSize is only called for Exprs where
+	// CEL does not know the size; EstimateSize is not called for literals defined inline in CEL.
+	EstimateSize(element Expr) *int64
+	// EstimateCallCost returns the estimated cost of an invocation, or nil if
+	// the estimator has no estimate to provide. EstimateCallCost is only called for functions
+	// that CEL does not already know the cost; EstimateCallCost is not called for standard operations
+	// or functions.
+	EstimateCallCost(name string, target *Expr, args []Expr) *CostEstimate
+}
+
+// Expr represents an AST node for the purpose of cost estimations.
+type Expr interface {
+	// Type returns the deduced type of the Expr.
+	Type() *exprpb.Type
+	// LiteralSize returns the size of the Expr if it is a literal defined inline in CEL, or
+	// nil if the Exprs size is not statically known.
+	LiteralSize() *int64
+}
+
+type expr struct {
+	t    *exprpb.Type
+	expr *exprpb.Expr
+}
+
+func (e expr) Type() *exprpb.Type {
+	return e.t
+}
+
+func (e expr) LiteralSize() *int64 {
+	var v int64
+	switch ek := e.expr.ExprKind.(type) {
+	case *exprpb.Expr_ConstExpr:
+		switch ck := ek.ConstExpr.ConstantKind.(type) {
+		case *exprpb.Constant_StringValue:
+			v = int64(len(ck.StringValue))
+		case *exprpb.Constant_BytesValue:
+			v = int64(len(ck.BytesValue))
+		default:
+			return nil
+		}
+	case *exprpb.Expr_ListExpr:
+		v = int64(len(ek.ListExpr.Elements))
+	case *exprpb.Expr_StructExpr:
+		if ek.StructExpr.MessageName == "" {
+			v = int64(len(ek.StructExpr.Entries))
+		}
+	default:
+		return nil
+	}
+
+	return &v
+}
+
+var (
+	identCost  = CostEstimate{Min: 1, Max: 1}
+	selectCost = CostEstimate{Min: 1, Max: 1}
+	constCost  = CostEstimate{Min: 0, Max: 0}
+)
+
+// CostEstimate represents an estimated cost range.
+type CostEstimate struct {
+	Min, Max int64
+}
+
+// Add adds the costs and returns the sum.
+func (ce CostEstimate) Add(rhs CostEstimate) CostEstimate {
+	return CostEstimate{
```

Comment:
Any thoughts on whether to use a pointer instead? It's fine to use a value, just curious though since in the `Expr` interface it's not clear whether the pointer is being used to make the return "optional", or if the pointer is used to indicate that the value is longer lived somewhow.

Reply:
A non-pointer seemed appropriate. It's just a composite scalar with operations that don't mutate state.